### DISCUSSION TITLE
Mouse Interrupt Integration

### DISCUSSION THREAD

**Discussion by RKennedy9064** [phil-opp/blog_os#795] [General]
I've been looking into integrating mouse input for my windowing library and it seems like its fairly similar to how the keyboard is setup. Would you be interested in an addendum to the post that also shows how to get mouse movement output to the screen. I'd love to help collaborate if that's something you're interested in, just let me know.

**Comment by phil-opp** [comment]
Yes, mouse input would be nice to have!

**Comment by RKennedy9064** [comment]
Awesome. I've been playing around with it and this is what I have for a rough draft if you want to test it out when you get a chance. There's probably some cleanup and optimizations that can be done. 

https://gist.github.com/RKennedy9064/5bffdfba096a1c4e3a1e10c50681be72

**Comment by phil-opp** [comment]
Looks good from a first glance. I think we could create a small crate for the `Mouse` type, similar to the `pc-keyboard` crate we already use. This way, we could add mouse support to the Hardware Interrupts in just a few paragraphs.

**Comment by RKennedy9064** [comment]
I like that idea. A lot of the initialization code is just reading and writing magic values that could be its own post. It looks like `mouse` is taken, but `pc-mouse` is open if you want to keep the name the same as `pc-keyboard`. Otherwise `ps2-mouse` is also open since it uses the ps2 port to interact with the mouse.

**Comment by phil-opp** [comment]
Yeah, I assumed that. I don't really have a strong preference about the name, but ps2-mouse sounds good.

**Comment by RKennedy9064** [comment]
I threw together a basic repo here https://github.com/rust-osdev/ps2-mouse. I added the crate to my kernel and it seems to be working correctly so far. I also added a basic example in the readme. Let me know if it works for you when you get a chance.

**Comment by RKennedy9064** [comment]
I also threw together a quick example of a basic implementation using the async/await posts keyboard handler here https://gist.github.com/RKennedy9064/de8000f8f63e14d9aa198240d7d9effb.
